UndeliverableException in stderr

### Steps to Reproduce

1. Download example application: `git clone https://github.com/VojtechMucha/micronaut-rxjava-stacktrace.git`
2. Execute test: `gw test`
3. Check console output for `io.reactivex.exceptions.UndeliverableException`

### Expected Behaviour

This (or very similar) issue was already reported as https://github.com/micronaut-projects/micronaut-core/issues/1799. As Graeme Rocher mentioned there: _It should never arrive to point of calling RxJavaPlugins.onError()_. But it currently happens.

### Actual Behaviour

Following exception is printed into stderr
```
    io.reactivex.exceptions.UndeliverableException: The exception could not be delivered to the consumer because it has already canceled/disposed the flow or the exception has nowhere to go to begin with. Further reading: https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error-handling | io.micronaut.http.client.exceptions.HttpClientException: Connect Error: connection timed out: www.google.com/172.217.23.196:81
        at io.reactivex.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:367)
        at io.reactivex.internal.operators.flowable.FlowableCreate$BaseEmitter.onError(FlowableCreate.java:275)
        at io.micronaut.http.client.DefaultHttpClient.lambda$null$27(DefaultHttpClient.java:1056)
        at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:577)
        at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:570)
        at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:549)
        at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:490)
        at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:615)
        at io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:608)
        at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:117)
        at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe$1.run(AbstractNioChannel.java:263)
        at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
        at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.micronaut.scheduling.instrument.InvocationInstrumenterWrappedRunnable.run(InvocationInstrumenterWrappedRunnable.java:48)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:834)
    Caused by: io.micronaut.http.client.exceptions.HttpClientException: Connect Error: connection timed out: www.google.com/172.217.23.196:81
        at io.micronaut.http.client.DefaultHttpClient.lambda$null$27(DefaultHttpClient.java:1057)
        ... 18 more
    Caused by: io.netty.channel.ConnectTimeoutException: connection timed out: www.google.com/172.217.23.196:81
        at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe$1.run(AbstractNioChannel.java:261)
        ... 10 more
    Exception in thread "nioEventLoopGroup-1-4" io.reactivex.exceptions.UndeliverableException: The exception could not be delivered to the consumer because it has already canceled/disposed the flow or the exception has nowhere to go to begin with. Further reading: https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error-handling | io.micronaut.http.client.exceptions.HttpClientException: Connect Error: connection timed out: www.google.com/172.217.23.196:81
        at io.reactivex.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:367)
        at io.reactivex.internal.operators.flowable.FlowableCreate$BaseEmitter.onError(FlowableCreate.java:275)
        at io.micronaut.http.client.DefaultHttpClient.lambda$null$27(DefaultHttpClient.java:1056)
        at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:577)
        at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:570)
        at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:549)
        at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:490)
        at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:615)
        at io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:608)
        at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:117)
        at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe$1.run(AbstractNioChannel.java:263)
        at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
        at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:500)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.micronaut.scheduling.instrument.InvocationInstrumenterWrappedRunnable.run(InvocationInstrumenterWrappedRunnable.java:48)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:834)
    Caused by: io.micronaut.http.client.exceptions.HttpClientException: Connect Error: connection timed out: www.google.com/172.217.23.196:81
        at io.micronaut.http.client.DefaultHttpClient.lambda$null$27(DefaultHttpClient.java:1057)
        ... 18 more
    Caused by: io.netty.channel.ConnectTimeoutException: connection timed out: www.google.com/172.217.23.196:81
        at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe$1.run(AbstractNioChannel.java:261)
        ... 10 more
```

### Environment Information

- **Operating System**: macOS 10.14.6
- **Micronaut Version:** 1.3.2
- **JDK Version:** openjdk version 11.0.2

### Example Application

https://github.com/VojtechMucha/micronaut-rxjava-stacktrace

--- a/http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java
+++ b/http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java
@@ -1112,7 +1112,6 @@ public class DefaultHttpClient implements
                                 } catch (Exception e) {
                                     emitter.onError(e);
                                 }
-
                             } else {
                                 Throwable cause = future.cause();
                                 emitter.onError(
@@ -1129,9 +1128,13 @@ public class DefaultHttpClient implements
                     addInstrumentedListener(connectionFuture, future -> {
                         if (!future.isSuccess()) {
                             Throwable cause = future.cause();
-                            emitter.onError(
-                                    new HttpClientException("Connect Error: " + cause.getMessage(), cause)
-                            );
+                            if (emitter.isCancelled()) {
+                                log.trace("Connection to {} failed, but emitter already cancelled.", requestURI, cause);
+                            } else {
+                                emitter.onError(
+                                        new HttpClientException("Connect Error: " + cause.getMessage(), cause)
+                                );
+                            }
                         } else {
                             try {
                                 sendRequestThroughChannel(

--- a/http-client/src/test/groovy/io/micronaut/http/client/ClientTimeoutSpec.groovy
+++ b/http-client/src/test/groovy/io/micronaut/http/client/ClientTimeoutSpec.groovy
@@ -1,0 +1,94 @@
+package io.micronaut.http.client
+
+
+import io.micronaut.context.annotation.ConfigurationProperties
+import io.micronaut.context.annotation.Property
+import io.micronaut.context.annotation.Requires
+import io.micronaut.http.annotation.Controller
+import io.micronaut.http.annotation.Get
+import io.micronaut.http.client.annotation.Client
+import io.micronaut.http.client.exceptions.HttpClientResponseException
+import io.micronaut.test.extensions.spock.annotation.MicronautTest
+import io.reactivex.Single
+import org.spockframework.runtime.IStandardStreamsListener
+import org.spockframework.runtime.StandardStreamsCapturer
+import spock.lang.Issue
+import spock.lang.Specification
+
+import javax.inject.Inject
+import java.time.Duration
+import java.util.concurrent.TimeUnit
+
+@Issue('https://github.com/micronaut-projects/micronaut-core/issues/2971')
+@MicronautTest
+@Property(name = 'spec.name', value = 'ClientTimeoutSpec')
+class ClientTimeoutSpec extends Specification {
+
+    @Inject
+    @Client('/')
+    RxHttpClient client
+
+    def 'test that UndeliverableException is not output to console'() {
+        setup:
+        def listener = new UndeliverableExceptionListener()
+        def capturer = new StandardStreamsCapturer()
+        capturer.addStandardStreamsListener(listener)
+        capturer.start()
+
+        when:
+        client.toBlocking().retrieve('/')
+
+        then:
+        def ex = thrown(HttpClientResponseException)
+        ex.message == 'Internal Server Error: The source did not signal an event for 1 seconds and has been terminated.'
+
+        and: 'wait for System.err messages to come in'
+        Thread.sleep(2000)
+        !listener.undeliverable
+
+        cleanup:
+        capturer.stop()
+        capturer.removeStandardStreamsListener(listener)
+    }
+}
+
+class UndeliverableExceptionListener implements IStandardStreamsListener {
+    boolean undeliverable = false
+
+    @Override
+    void standardOut(String message) { }
+
+    @Override
+    void standardErr(String message) {
+        undeliverable = undeliverable || message.contains('io.reactivex.exceptions.UndeliverableException')
+    }
+}
+
+@Requires(property = 'spec.name', value = 'ClientTimeoutSpec')
+@Controller('/')
+class TimeoutController {
+    @Inject
+    TimeoutClient client
+
+    @Get
+    Single<String> get() {
+        // Client will timeout in 2s (see TimeoutClientConfiguration below); we timeout the stream in 1s.
+        client.get().timeout(1, TimeUnit.SECONDS)
+    }
+}
+
+@Requires(property = 'spec.name', value = 'ClientTimeoutSpec')
+@Requires(beans = TimeoutClientConfiguration)
+@Client(value = 'http://www.google.com:81/', configuration = TimeoutClientConfiguration)
+interface TimeoutClient {
+    @Get
+    Single<String> get();
+}
+
+@ConfigurationProperties(PREFIX)
+class TimeoutClientConfiguration extends DefaultHttpClientConfiguration {
+    @Override
+    Optional<Duration> getConnectTimeout() {
+        Optional.of(Duration.ofSeconds(2))
+    }
+}

